### DISCUSSION TITLE
Use environment variables in pactumjs

### DISCUSSION THREAD

**Discussion by vladparl** [pactumjs/pactum#122] [Q&A] [answered]
Hello,

I am using pactum-cucumber-boilerplate project and I have following scenario:

Given base URL is "some URL"
When make a POST request with body to /auth/login
        """
        {
            "username": "user@user.com",
            "password": "Password!"
        }       
        """

Instead of using username and password directly in gherkin json body, I am wondering is there any way to save them somewhere in file and call them here in Gherkin scenario?

**Comment by ASaiAnudeep** [comment, marked as answer]
We can use [DataMaps](https://pactumjs.github.io/#/data-management?id=data-map) to make common data available across test cases.

In global hooks:

```js
const { stash } = require('pactum');
const { Before } = require('@cucumber/cucumber');

Before(() => {
  stash.addDataMap({
    'App': {
      'Username': process.env.APP_USERNAME || 'default-value',
      'Password': process.env.APP_PASSWORD || 'default-value'
    }
  });
});
```

In feature file:

```gherkin
Given base URL is "some URL"
When make a POST request with body to /auth/login
"""
{
  "username": "$M{App.Username}",
  "password": "$M{App.Password}"
}
"""
```

**Reply by vladparl** [reply]
What if for example want to save another value in DataMap?

I tried with  'Url': process.env.APP_URL || 'https://some_value'

In feature file:
Given base URL is "$M{App.Url}"

... and it returned error message "Invalid Url" set something like that.

**Comment by vladparl** [comment]
Thank you ASaiAnudeep, problem solved. Works like a charm 👍
